r/R/dplyr.R

```diff
@@ -619,9 +619,23 @@ collect.arrow_dplyr_query <- function(x, as_data_frame = TRUE, ...) {
     restore_dplyr_features(tab, x)
   }
 }
-collect.ArrowTabular <- as.data.frame.ArrowTabular
+collect.ArrowTabular <- function(x, as_data_frame = TRUE, ...) {
+  if (as_data_frame) as.data.frame(x, ...) else x
+}
 collect.Dataset <- function(x, ...) dplyr::collect(arrow_dplyr_query(x), ...)
 
+compute.arrow_dplyr_query <- function(x, ...) dplyr::collect(x, as_data_frame = FALSE)
+compute.ArrowTabular <- function(x, ...) x
+compute.Dataset <- function(x, ...) {
```

Comment:
This is nice and concise. Another way you could do this would be to factor out the `as_data_frame = FALSE` section of `collect()` and call that `compute()`, and then you'd have

```
collect <- function(x, as_data_frame = TRUE, ...) {
  tab <- dplyr::compute(x, ...) 
    if (as_data_frame) {
      df <- as.data.frame(tab)
      tab$invalidate()
      restore_dplyr_features(df, x)
    } else {
      restore_dplyr_features(tab, x)
    }
}
```

I think this version is slightly better because it lets us have shorter functions that compose together logically.

Reply:
I thought of doing this, but didn't at first because we want `compute()` to call `restore_dplyr_features()` too, and I wasn't sure if that is fast and idempotent. I'll check.

Reply:
Not to scope creep, but now that we can project with renaming etc. (`mutate()`), I think all `restore_dplyr_features()` needs to do is keep track of `group_by_vars`. There might be a simpler way to handle that that solves your concern.

Reply:
I spent about an hour trying various different things here, such as paring back what `restore_dplyr_features()` does and also factoring out the compute code into a separate internal function and calling it from both `collect` and `compute`, but this caused test failures. Investigating these seems beyond the scope of what we're trying to achieve here and I think my time would be better spent on other work. Can we leave this as is for now and open a Jira for improvement later?

Reply:
Sure, sounds like a plan

Reply:
ARROW-12282